ndb/src/google/cloud/ndb/model.py

```diff
@@ -1407,10 +1409,212 @@ def __init__(self, *args, **kwargs):
+        default (bytes): The default value for this property.
+        choices (Iterable[bytes]): A container of allowed values for this
+            property.
+        validator (Callable[[Property, Any], bool]): A validator to be used
+            to check values.
+        verbose_name (str): A longer, user-friendly name for this property.
+        write_empty_list (bool): Indicates if an empty list should be written
+            to the datastore.
+
+    Raises:
+        NotImplementedError: If the property is both compressed and indexed.
+    """
+
+    _indexed = False
+    _compressed = False
+
+    def __init__(
+        self,
+        name=None,
+        compressed=False,
+        *,
+        indexed=None,
+        repeated=None,
+        required=None,
+        default=None,
+        choices=None,
+        validator=None,
+        verbose_name=None,
+        write_empty_list=None
+    ):
+        super(BlobProperty, self).__init__(
+            name=name,
+            indexed=indexed,
+            repeated=repeated,
+            required=required,
+            default=default,
+            choices=choices,
+            validator=validator,
+            verbose_name=verbose_name,
+            write_empty_list=write_empty_list,
+        )
+        self._compressed = compressed
+        if compressed and self._indexed:
+            raise NotImplementedError(
+                "BlobProperty {} cannot be compressed and "
+                "indexed at the same time.".format(self._name)
+            )
+
+    def _value_to_repr(self, value):
+        """Turn the value into a user friendly representation.
+
+        .. note::
+
+            This will truncate the value based on the "visual" length, e.g.
+            if it contains many ``\\xXX`` or ``\\uUUUU`` sequences, those
+            will count against the length as more than one character.
+
+        Args:
+            value (Any): The value to convert to a pretty-print ``repr``.
+
+        Returns:
+            str: The ``repr`` of the "true" value.
+        """
+        long_repr = super(BlobProperty, self)._value_to_repr(value)
+        if len(long_repr) > _MAX_STRING_LENGTH + 4:
+            # Truncate, assuming the final character is the closing quote.
+            long_repr = long_repr[:_MAX_STRING_LENGTH] + "..." + long_repr[-1]
+        return long_repr
+
+    def _validate(self, value):
+        """Validate a ``value`` before setting it.
+
+        Args:
+            value (bytes): The value to check.
+
+        Raises:
+            .BadValueError: If ``value`` is not a :class:`bytes`.
+            .BadValueError: If the current property is indexed but the value
+                exceeds the maximum length (1500 bytes).
+        """
+        if not isinstance(value, bytes):
+            raise exceptions.BadValueError(
+                "Expected bytes, got {!r}".format(value)
+            )
+
+        if (
+            self._indexed
+            and not isinstance(self, TextProperty)
```

Comment:
This should probably be moved into `TextProperty._validate` for special handling. I think it's bad form for a base class to refer to a child class in this way.

Reply:
Yep.

Reply:
So I dug a little deeper on this and realized the issue. The `Property._call_to_base_type` and `Property._call_shallow_validation` methods will call **every** `_validate` method on the MRO, so this will get invoked for `TextProperty` instances **even though** `TextProperty._validate` is implemented (and does not invoke `super()`).

Reply:
Right, but `TextProperty._validate` should get called before this one, and if it raises `BadValueError` this won't get called, right? Still seems like `TextProperty._validate` is still the right place for it.

Reply:
Not quite. What's happening here is an error for "index properties with greater than 1500 bytes, except for a text property". So it's a custom **exemption** for `TextProperty`.

I could rely on an attribute like `self._LENGTH_EXEMPT` and then just define `BlobProperty._LENGTH_EXEMPT = False` and `TextProperty._LENGTH_EXEMPT = True`.

WDYT?

Reply:
Ah right. Or mabye just make _MAX_STRING_LENGTH a class level attribute and have TextProperty override it with `0` or `None` meaning no limit. Either way.

Reply:
OK this is extra fun. I have no idea why it's in the [original][1] because `TextProperty._validate` also [checks][2] the length.

I'm just going to remove the type check here because it's equivalent to the branch of `TextProperty._validate` when `value` is a `bytes` instance (except for the decode to UTF-8 check).

[1]: https://github.com/GoogleCloudPlatform/datastore-ndb-python/blob/cf4cab3f1f69cd04e1a9229871be466b53729f3f/ndb/model.py#L1715-L1720
[2]: https://github.com/GoogleCloudPlatform/datastore-ndb-python/blob/cf4cab3f1f69cd04e1a9229871be466b53729f3f/ndb/model.py#L1782-L1785

Reply:
Ha, ok.